### ISSUE TITLE
Live training dashboard in the web demo

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#264]
## Summary
Extend the web demo with a live training dashboard. Roadmap Phase 4 item (the only remaining unchecked Phase 4 box). Current demos at https://rjwalters.github.io/thrust/ are inference-only (pre-trained policies).

## Motivation
A "watch it learn" view — reward/loss curves updating during training — is the highest-visibility showcase of the train-in-Rust story and complements the existing inference demos.

## Scope / pointers
- Frontend: `web/src/pages/` (React + Vite; existing pages: CartPole, Snake, Pong, bandit).
- Examples already emit learning curves via the `CURVE_CSV` env var — a natural data source, or stream metrics from a WASM training loop.
- Decide: replay a recorded training run (simple) vs. live in-browser training via the WASM bindings (`src/wasm.rs`, harder).

## Acceptance criteria
- [ ] A dashboard page showing reward/loss over training steps for at least one env.
- [ ] Wired into the site nav (`web/src/App.tsx`) and deploys via the existing Pages workflow.
- [ ] Document the data source (recorded CSV vs. live WASM training).

## Notes
Recommend starting with replay-of-recorded-run to ship value fast, then consider live in-browser training as a follow-up.

**Comment by rjwalters** [comment]
## Curator Enhancement

Curating this issue per user request. Context and guidance added below.

---

## Technical Context

### Current state of the web demo

The site at https://rjwalters.github.io/thrust/ serves five pages: Home, CartPole, Snake, Pong, SimpleBandit. All are **inference-only** — they load pre-trained model weights (committed as JSON files under `web/public/`) and run the policy in WASM. No training loop runs in-browser today.

### WASM binding scope today

`src/wasm.rs` exposes environment step/reset/policy-load primitives for each game. There is **no training-loop WASM binding** (no `WasmTrainer`, no gradient-update export). Adding live in-browser training would require non-trivial new WASM surface and is explicitly out of scope for this issue (the issue itself recommends starting with replay-of-recorded-run).

### CURVE_CSV mechanism — data source

Six training examples already emit a learning-curve CSV when `CURVE_CSV=<path>` is set. They all share the same two-column schema:

```
env_steps,mean_episode_reward
1000,23.4500
2000,45.1200
...
```

Header is always `env_steps,mean_episode_reward`. Rows are emitted once per logging interval. Examples that support this (all verified in-repo):

| Example | Algorithm |
|---|---|
| `examples/games/cartpole/train_cartpole_dqn.rs` | DQN |
| `examples/games/cartpole/train_cartpole_a2c.rs` | A2C |
| `examples/games/cartpole/train_cartpole_modern.rs` | PPO |
| `examples/games/pendulum/train_sac.rs` | SAC |
| `examples/games/mountain_car/train_sac_mountain_car.rs` | SAC |
| `examples/games/grid_world/train_dqn_grid_world.rs` | DQN |

The three CartPole examples (A2C, DQN, PPO) share the **same environment seed and step budget**, so their curves overlay directly — a compelling multi-algorithm comparison chart is available out of the box.

### Recommended data flow: replay-of-recorded-run

Commit one or more CSV files to `web/public/` (same pattern as the model `.json` files). The `deploy-pages.yml` workflow runs `pnpm build` which packages everything under `web/public/` — no workflow changes needed.

**To generate**:
```bash
# CartPole: A2C, DQN, PPO (same seed, same 60 k step budget)
CURVE_CSV=web/public/cartpole_curves_a2c.csv cargo run --example train_cartpole_a2c --features training --release
CURVE_CSV=web/public/cartpole_curves_dqn.csv cargo run --example train_cartpole_dqn --features training --release
CURVE_CSV=web/public/cartpole_curves_ppo.csv cargo run --example train_cartpole_modern --features training --release
```

The page then fetches these at runtime via `fetch(`${import.meta.env.BASE_URL}cartpole_curves_dqn.csv`)` (mirroring how `CartPolePage.tsx` fetches `cartpole_model_best.json`).

### Charting approach

No chart library is currently installed. The existing stack is React 19, Vite, Tailwind. Recommended addition: **`recharts`** (pure-React, SVG-based, zero canvas setup). The `LineChart` component with multiple `<Line>` series handles the A2C/DQN/PPO overlay naturally.

Alternative (zero-dependency): a hand-rolled SVG polyline using `<svg viewBox>` — viable but more implementation work for a first pass.

### Page structure and nav wiring

Existing pages follow `GamePageLayout` (title + visualization + controls + gameDynamics + neuralNetworkArchitecture sections). The dashboard page can use a **simpler custom layout** — just a chart and a prose explanation — since there is no interactive WASM control panel.

Nav wiring pattern from `App.tsx`:
```tsx
<Route path="/dashboard" element={<TrainingDashboardPage />} />
```

Home page: add a new `<Link to="/dashboard">` card in the same `md:grid-cols-2` grid.

---

## Implementation Guidance

### Recommended scope for the Builder

**Phase 1 (this issue)**: CartPole only, replay-of-recorded-run, three algorithms overlaid on one chart.

1. Generate the three CartPole CSVs locally and commit them to `web/public/`.
2. Add `recharts` to `web/package.json`.
3. Create `web/src/pages/TrainingDashboardPage.tsx`:
   - `useEffect` to `fetch` each CSV on mount, parse with `String.split('\n')` and map rows to `{env_steps, value}` objects (same pattern as the `modelInfo` fetch in CartPolePage)
   - `recharts` `<LineChart>` with three `<Line>` series (A2C, DQN, PPO), an `<XAxis>` labelled "env steps", a `<YAxis>` labelled "mean episode reward", and a `<Legend>`
   - A prose section below explaining the three algorithms and what the convergence curves show
4. Wire the route in `web/src/App.tsx`.
5. Add a "Training Dashboard" card in `web/src/pages/Home.tsx`.
6. Add a Playwright smoke test in `web/tests/site-loads.spec.ts`.

**Out of scope** (follow-up): live in-browser WASM training, loss curves (not emitted by current examples), additional environments.

---

## Acceptance Criteria

- [ ] A `/dashboard` (or `/training`) page is accessible from the site nav (Home card + route in App.tsx).
- [ ] The page renders a reward curve chart for at least one environment (CartPole recommended), showing `env_steps` on the x-axis and `mean_episode_reward` on the y-axis.
- [ ] At least one algorithm series is displayed; overlaying A2C + DQN + PPO is the stretch goal.
- [ ] The data source (CSV file, recorded run) is documented on the page (e.g. a sentence noting "Replay of a recorded training run — 60 k env steps, CartPole-v1, seed 0").
- [ ] The page is wired into `web/src/App.tsx` with a `<Route>` and linked from `web/src/pages/Home.tsx`.
- [ ] `pnpm build` passes without errors or TypeScript warnings.
- [ ] The page loads cleanly in the Playwright smoke suite (`pnpm test`).
- [ ] The deployment pipeline (`deploy-pages.yml`) picks up the new page automatically — no workflow changes required.

---

## Affected Files

**Existing files to modify** (all confirmed on `origin/main`):

- `web/src/App.tsx` — add `<Route path="/dashboard" element={<TrainingDashboardPage />} />` and import
- `web/src/pages/Home.tsx` — add dashboard card to the grid
- `web/package.json` — add `recharts` dependency (e.g. `"recharts": "^2.15.0"`)
- `web/pnpm-lock.yaml` — updated automatically by `pnpm install`
- `web/tests/site-loads.spec.ts` — add a smoke test for the dashboard page

**New files to create** (do not exist on `origin/main` — Builder creates them):

- `web/src/pages/TrainingDashboardPage.tsx` — new page component
- `web/public/cartpole_curves_a2c.csv` — generated by `train_cartpole_a2c` example, committed
- `web/public/cartpole_curves_dqn.csv` — generated by `train_cartpole_dqn` example, committed
- `web/public/cartpole_curves_ppo.csv` — generated by `train_cartpole_modern` example, committed

Scope guard: the Builder should not modify `src/wasm.rs`, any `train_*.rs` examples, or the CI workflow. This is a purely frontend + static-asset change.

---

## Test Plan

- [ ] Manual: `cd web && pnpm dev`, navigate to `/thrust/dashboard`, verify the chart renders with at least one visible curve and correct axis labels.
- [ ] Manual: check the browser console for errors (no failed fetches, no WASM errors — the page does not load WASM).
- [ ] Manual: verify the Home page card links to the dashboard page.
- [ ] Build: `cd web && pnpm build` completes without TypeScript or Vite errors.
- [ ] Lint: `cd web && pnpm lint` passes (Biome + ESLint configured in the project).
- [ ] Playwright: `cd web && pnpm test` includes a new `"training dashboard page loads"` test that navigates to the route and asserts the `<h1>` contains "Training" (or equivalent), no page errors.
- [ ] Deploy: push to `main` and verify GitHub Pages deploys successfully via the Actions tab; confirm the live URL `/thrust/dashboard` renders the chart.

---

## Dependencies

No blocking issues. All prerequisites are already shipped:
- WASM build + Pages deploy pipeline: in place
- CSV-emitting training examples: in place (six examples, all confirmed)
- React + Vite + Tailwind stack: in place
